external-crates/move/crates/move-compiler/src/parser/syntax.rs

```diff
@@ -2965,7 +2936,154 @@ fn parse_module(
+                    // if old syntax is used
+                    if context.tokens.lookahead2() == Ok((Tok::Identifier, Tok::LBrace)) {
+                        return Err((
+                            unexpected_token_error(
+                                context.tokens,
+                                "only 'spec', drop the 'fun' keyword",
+                            ),
+                            None,
+                        ));
+                    }
+                    Ok(ModuleMember::Spec(
+                        singleton_module_spec_block(
+                            context,
+                            start_loc,
+                            attributes,
+                            parse_spec_function,
+                        )
+                        .map_err(|e| (e, None))?,
+                    ))
+                }
+                _ => {
+                    // Regular spec block
+                    Ok(ModuleMember::Spec(
+                        parse_spec_block(attributes, context).map_err(|e| (e, None))?,
+                    ))
+                }
+            }
+        }
+        // Regular move constructs
+        Tok::Friend => Ok(ModuleMember::Friend(
+            parse_friend_decl(attributes, context).map_err(|e| (e, None))?,
+        )),
+        _ => {
+            context.tokens.match_doc_comments();
+            let start_loc = context.tokens.start_loc();
+            let modifiers = parse_module_member_modifiers(context).map_err(|e| (e, None))?;
+            match context.tokens.peek() {
+                Tok::Const => Ok(ModuleMember::Constant(
+                    parse_constant_decl(attributes, start_loc, modifiers, context)
+                        .map_err(|e| (e, None))?,
+                )),
+                Tok::Fun => Ok(ModuleMember::Function(
+                    parse_function_decl(attributes, start_loc, modifiers, context)
+                        .map_err(|e| (e, None))?,
+                )),
+                Tok::Struct => Ok(ModuleMember::Struct(
+                    parse_struct_decl(attributes, start_loc, modifiers, context)
+                        .map_err(|e| (e, None))?,
+                )),
+                Tok::Use => Ok(ModuleMember::Use(
+                    parse_use_decl(attributes, start_loc, modifiers, context)
+                        .map_err(|e| (e, None))?,
+                )),
+                _ => Err((
+                    unexpected_token_error(
+                        context.tokens,
+                        &format!(
+                            "a module member: '{}', '{}', '{}', '{}', '{}', or '{}'",
+                            Tok::Spec,
+                            Tok::Use,
+                            Tok::Friend,
+                            Tok::Const,
+                            Tok::Fun,
+                            Tok::Struct
+                        ),
+                    ),
+                    Some(attributes),
```

Comment:
Could we just explicitly check for `Tok::Module` to handle this instead? 
We can then enumerate the error cases with an enum, e.g. 
```
enum ErrCase {
  Unknown, // diag logged in env
  ContinueToModule(Option<Attributes>)
}
```

Reply:
I don't want to put diagnostic into env prematurely for the reason I explained in the other [comment](https://github.com/MystenLabs/sui/pull/15794#discussion_r1461197065). I am also not sure how much cleaner the code would be if we did that...

Reply:
> I don't want to put diagnostic into env prematurely for the reason I explained in the other comment 

Not sure I follow. Either you've already set the error, and can disregard it when logging it. Or its the first error and you want to add it. In either case you don't need to hold onto it locally 

> I am also not sure how much cleaner the code would be if we did that...

It is going to be just about the same as the Option, but more amenable to future changes with some intent signaling too for future changes. Its nice to know that the attributes are Some iff there was an error

Reply:
The thing is that encountering `Tok::Module` here may actually be first legit parsing error and I want to return it as-is, so that the enclosing function can record it with `set_parsing_error`. Otherwise, if we treat encountering `Tok::Module` here as some special error, what kind of diagnostic would we record? Clearly, we can simply insert the same unexpected token error in the enclosing function, but that would create a redundancy of sorts.

I am likely misunderstanding something here, though, as in either case we need to pass other diag to the enclosing function so return type `Result<ModuleMember, ErrCase>` would not work here, if that return type was the intention behind introducing the `ErrCase` enum... Or was the intent to have `Result<ModuleMember, (Box<Diagnostic>, ErrCase>)>` (in which case we still need to generate the unexpected token error for `Tok::Module`...